### ISSUE TITLE
Custom resources `include_recipe` expanded run_list ordering problem

### ISSUE THREAD

**Issue by maoueh** [chef/chef#4260]
Chef client 12.5, CentOS 6.6, x86_64

When a custom resource (`root_prepare`) (be it Chef 12.5 resource or `LWRP`) includes a recipe (`root::bootstrap`) that in turn include a recipe (`common::default`) that is already part of the execution list via another included recipe (`app::default`), ordering is wrong and initial `include_recipe 'common::default'` is executed too late.

This is a hell of a sentence so let's stick with an example so everything is clear (examples available in `gist` form https://gist.github.com/maoueh/f85c30d7869294d928c4 and an archive could be provided on demand):

```ruby
# cookbooks/common/recipes/default.rb
log 'Should log first (in common::default).'
```

```ruby
# cookbooks/app/recipes/default.rb
include_recipe 'common::default'
```

```ruby
# cookbooks/root/resources/prepare.rb
actions :execute
default_action :execute

attribute :name, :kind_of => String, :name_attribute => true
```

```ruby
# cookbooks/root/providers/prepare.rb
include Chef::DSL::IncludeRecipe

action :execute do
  log "In LWRP, before include_recipe 'root::bootstrap'"
  include_recipe 'root::bootstrap'
  log "In LWRP, after include_recipe 'root::bootstrap'"
end
```

```ruby
# cookbooks/root/recipes/bootstrap.rb
include_recipe 'common::default'

log 'Should log second (in root::boostrap).'
```

```ruby
# cookbooks/root/recipes/default.rb
root_prepare 'id'

include_recipe 'app::default'

log 'Should log third (in root::default).'
```

Running this using `chef-client -z -o 'root::default'` yields the following output:

```
[root@host sample]# chef-client -z -o 'root'
[2015-12-09T16:13:13+00:00] WARN: No config file found or specified on command line, using command line options.
Starting Chef Client, version 12.4.3
[2015-12-09T16:13:15+00:00] WARN: Run List override has been provided.
[2015-12-09T16:13:15+00:00] WARN: Original Run List: []
[2015-12-09T16:13:15+00:00] WARN: Overridden Run List: [recipe[root]]
resolving cookbooks for run list: ["root"]
Synchronizing Cookbooks:
  - common
  - app
  - root
Compiling Cookbooks...
Converging 3 resources
Recipe: root::default
  * root_prepare[id] action execute (up to date)
  * log[In LWRP, before include_recipe 'root::bootstrap'] action write

Recipe: root::bootstrap
  * log[Should log second (in root::boostrap).] action write

  * log[In LWRP, after include_recipe 'root::bootstrap'] action write

Recipe: common::default
  * log[Should log first (in common::default).] action write

Recipe: root::default
  * log[Should log third (in root::default).] action write

[2015-12-09T16:13:15+00:00] WARN: Skipping final node save because override_runlist was given

Running handlers:
Running handlers complete
Chef Client finished, 5/6 resources updated in 2.154498603 seconds
```

As we can see, the `root_prepare` resource is executed and the `root::bootstrap` recipe is correctly run in the right order. However, when executing `root::bootstrap` resources, its `include_recipe 'common::default'` is not yet executed meaning the resources in this recipe are also not executed. Instead, they are executed only later as part of the `include_recipe 'app::default'` which also includes the `common::default` recipe. 

This is problematic when the included recipe is for example `yum-epel` and that a package from the `epel` repository is installed inside `root::bootstrap` included recipe. The bug causes the package to be not found and chef execution run fails.

I tested this with chef client `11.18`, `12.2`, `12.3`, `12.4` and `12.5`. And they all failed. This is present for so long that maybe I'm missing something and current behavior is correct. But from my knowledge of how ordering should work, that's an error.

Current workaround is to actually do the `include_recipe` outside of my custom resource so order is correct.

Hope all is clear. Don't hesitate to ask further details if something is not.

Regards,
Matt

P.S.

Expected expansion phases

```
# From root::default (run_list)
|root_prepare 'id'
|include_recipe 'app::default'
|log 'Should log third (in root::default).'

--------------------------------------------------------

# From root_prepare 'id'
|log "In LWRP, before include_recipe 'root::bootstrap'"
|include_recipe 'root::bootstrap'
|log "In LWRP, after include_recipe 'root::bootstrap'"

# From include_recipe 'app::default'
|include_recipe 'common::default'

log 'Should log third (in root::default).'

--------------------------------------------------------

log "In LWRP, before include_recipe 'root::bootstrap'"
# From root::bootstrap
|include_recipe 'common::default'
|log 'Should log second (in root::boostrap).
log "In LWRP, after include_recipe 'root::bootstrap'"

# From include_recipe 'app::default'
|log 'Should log first (in common::default).'

log 'Should log third (in root::default).'

--------------------------------------------------------

log "In LWRP, before include_recipe 'root::bootstrap'"
# From include_recipe 'common::default'
|log 'Should log first (in common::default).'
log 'Should log second (in root::boostrap).
log "In LWRP, after include_recipe 'root::bootstrap'"
log 'Should log first (in common::default).' <-- Should be removed since resource already present
log 'Should log third (in root::default).'

--------------------------------------------------------

log "In LWRP, before include_recipe 'root::bootstrap'"
log 'Should log first (in common::default).'
log 'Should log second (in root::boostrap).
log "In LWRP, after include_recipe 'root::bootstrap'"
log 'Should log third (in root::default).'
```

Actual expansion phases:

```
... (same as before until here)

log "In LWRP, before include_recipe 'root::bootstrap'"
# From include_recipe 'common::default'
|log 'Should log first (in common::default).'  <-- Actually removed even if before second resource
log 'Should log second (in root::boostrap).
log "In LWRP, after include_recipe 'root::bootstrap'"
log 'Should log first (in common::default).' 
log 'Should log third (in root::default).'

--------------------------------------------------------

log "In LWRP, before include_recipe 'root::bootstrap'"
log 'Should log second (in root::boostrap).
log "In LWRP, after include_recipe 'root::bootstrap'"
log 'Should log first (in common::default).'
log 'Should log third (in root::default).'
```

**Comment by lamont-granquist** [comment]
LWRPs/resources execute in a sub-run_context/resource_collection all of their own.  That context isn't even created until the LWRP provider executes in the _converge_ phase of the parent run_context.  

So execution order:

- all the root run_context compile phases occur, no LWRPs providers are executed, resource code runs (validation that things that should be Strings are strings, etc), all include_recipes / chef_gems / ruby require lines (BUT DONT DO THIS) / etc all run now
- root compile phase starts
- first LWRP that this is encountered creates its own run_context
   - compile phase of the LWRP runs running all the ruby code / include_recipe / require / constructing resource collection
   - converge phase of the LWRP runs running all the non-LWRP resources
   - last_updated_by_state bubbles up to the parent resource
- root compile phase continues....

This has the beneficial effect that all pure-ruby code that occurs in the LWRP provider is actually lazied to the execution of the provider in the compile phase of the root run_context.

That is beneficial because now you can do a `chef_gem "whatever"` in the root run_context and that'll install the gem at converge time -- but later LWRPs can `require 'whatever'` in their providers and that'll lazily run afterwards.  That avoids the race we used to have when ruby require lines in the root run_context would all run before chef_gem's could ever run and we wound up forcing chef_gem to compile time and that created an arms race of garbage (build-essentials, apt, libxml2, zlib, etc) that had to be forced to install at compile time).

Anyway, pretty sure you just discovered how the execution model actually works, and that last paragraph there holds the use case that would cause the world to implode if we tried to change the execution model.

So, TL;DR:  pretty sure NOTABUG

**Comment by maoueh** [comment]
@lamont-granquist Thanks for the clarifications, very appreciate. Getting insights on how Chef internals run exactly is always a good things when debugging hard to understand problems.

Ok, so following this, and if my understanding is right, actual expansion phase looks more like this (I modified slightly the example so my real use case is shown):

```
root_prepare 'id'
yum 'epel' do ... end

-----------------------------------------------------
[New Context]
# From root_prepare 'id'
|include_recipe 'yum-epel'
|package 'yasm'
[/New Context]
yum 'epel' do ... end

-----------------------------------------------------
[New Context]
# From include_recipe 'yum-epel'
|yum 'epel' do ... end <-- Removed
package 'yasm'
[/New Context]
yum 'epel' do ... end
```

Order seems right, the bug to me seems more to be that the `yum 'epel'` resource is not executed as part of the `New Context` run. This is problematic because there other next resource (`package 'yasm'`) relies on the fact that the `yum 'epel'` have been executed at this point. 

The question is, why `yum 'epel' do ... end` resource is being excluded from the `New Context` run. I guess is because the parent context already has the resource. But the resource will be executed only later and I think it's not the right decision to remove the one in the `New Context` run.

**Comment by lamont-granquist** [comment]
include_recipe 'yum-epel' is only every 'run' once.  it isn't like a method call.   if you're trying to call it twice and expecting it to execute twice that is 'smell' that you want a resource instead of a recipe.

**Comment by lamont-granquist** [comment]
And generally include_recipe from a provider class tends to end badly.  Either the include_recipe should be moved into recipe code, or else the recipe needs to be converted to a resource of its own.  Its kind of a code smell on its own.  Its there to use if you can avoid the sharp edges, but you just discovered all the sharp edges and why its not a very good pattern to use.

**Comment by maoueh** [comment]
Don't get my wrong, I don't expect the `yum` resource to ran twice, I really expect to run it only once. What I was expecting is that the one from the provider would run and not the second one since the one in the provider "seen" first.

So, if my custom resource needs to install a package that depends on the `yum-epel` repository, proposed solution would be to force all users of my resource to `include_recipe 'yum-epel'` outside the resource:

```
include_recipe 'yum-epel'
root_prepare 'id'
```

Don't you find this a little bit limiting? It's fine by me because my custom resources are used internally in the same cookbook. Main drawback is that it exposes implementation details of my custom resource and increases duplication.

I'm ok if this is a won't fix.

**Comment by lamont-granquist** [comment]
So one thing I can see from your code above is that "include_recipe" does not open a new context, only LWRPs do.  So this bit does not read correct:

```
[New Context]
# From include_recipe 'yum-epel'
```

**Comment by maoueh** [comment]
This was coming from the `root_prepare 'id'` which opens a new context right?:

```
[Parent Context]
root_prepare 'id'
yum 'epel' do ... end
[/Parent Context]

expands to

[Parent Context]
[Child Context]
include_recipe 'yum-epel'
package 'yasm'
[/Child Context]
yum 'epel' do ... end
[/Parent Context]
```

**Comment by lamont-granquist** [comment]
Okay so lets rewind up to your example at the very top.

When chef outputs "Compiling Cookbooks..." the work that has been done is to parse all of your recipes in the top context:

```
# cookbooks/root/recipes/default.rb
root_prepare 'id'

include_recipe 'app::default'

log 'Should log third (in root::default).'
```

```
# cookbooks/app/recipes/default.rb
include_recipe 'common::default'
```

```
# cookbooks/common/recipes/default.rb
log 'Should log first (in common::default).'
```

This constructs (but does not yet execute the resource collection):

```
root_prepare 'id'
log 'Should log first (in common::default).'
log 'Should log third (in root::default).'
```

It also marks those recipes as 'seen' so that they will never be executed again:

* root::default
* app::default
* common::default

We're now down at "Converging 3 resources" in the chef output.

The next lines enter the "root_prepare[id]" provider, this code is what get executed:

```
# cookbooks/root/providers/prepare.rb
include Chef::DSL::IncludeRecipe

action :execute do
  log "In LWRP, before include_recipe 'root::bootstrap'"
  include_recipe 'root::bootstrap'
  log "In LWRP, after include_recipe 'root::bootstrap'"
end
```

```
# cookbooks/root/recipes/bootstrap.rb
include_recipe 'common::default'

log 'Should log second (in root::boostrap).'
```

This will:

1.  place the 'before' log line on the sub-resource_collection
2.  execute the include_recipe root::bootstrap since that recipe has not been SEEN yet
3.  will skip include_recipe common::default since that has been SEEN by the parent context already (the include_recipe code literally looks at a hash and "common::default" is already there, so it no-ops the whole thing)
4.  will add the "second" log line on the sub-resource_collection
5.  will add the "after" log line on the sub-resource_collection

So you get:

```
log "In LWRP, before include_recipe 'root::bootstrap'"
log 'Should log second (in root::boostrap).'
log "In LWRP, after include_recipe 'root::bootstrap'"
```

Which then converges and you get those lines output.

Then it pops back up to the parent run context, and you get the "first" log lines and "third" log lines output.

So there's only one sub-context created by the root_prepare resource/provider.  Both the compile and converge phases of that provider are trigger late in the parent run_context's converge phase.  The include_recipes are only ever parsed once, and they are compile phase, so they're early so any include_recipe in any top-level recipe code always 'wins' over any include_recipe inside of any provider.

And yeah, this means that you can't express "this resource requires this recipe to run" in a perfectly sane manner with this mechanism.  You probably want to either simply document that the user must include the recipe before using your resource -- or else try to convert the recipe to a resource (since resources can be run twice).

**Comment by aspyatkin** [comment]
Sorry to interfere, I've run into the similar problem.
I have a [wrapper cookbook](https://github.com/aspyatkin/latest-git) to install `git` on the development virtual machine. I've tweaked some attributes to build latest `git` from source (obviously it'll need `build-essential` cookbook at first).
So my understanding is that when I add `recipe[latest-git]` to node's run list, Chef will at first install `build-essential` packages, then will try to build `git`. However, this doesn't seem to happen. I've tried running my recipe on several machines (Ubuntu 14.04 server fresh setup on AWS), and it fails unless you add `build-essential` recipe to the node's run list.
I've looked into the code of `git` cookbook. Internally, `git::source` recipe makes use of custom resource named `git_client`. Inside the resource provider, `build-essential` is included.

I wonder whether I am doing the wrapper cookbooks wrong? I want to simplify things so when I add wrapper cookbook recipes to run_list, all dependencies get installed before they are needed e.g. to build something from source. This is much annoying when you have to look through the other cookbook requirements and add them to your cookbook recipes.

**Comment by lamont-granquist** [comment]
@aspyatkin you should do this explicitly yourself:

```
include_recipe "build-essential"

git_client "...stuff..." do
  [....stuff...]
end
````

and make your wrapper cookbook depend on build-essentials and git.

relying on the LWRP to include_recipe is going to be problematic like I outlined.

**Comment by lamont-granquist** [comment]
There doesn't seem to be any bugs here.

Because of compile/converge mode include_recipe in LWRPs is going to be inherently difficult to understand.

If it fails to work, the best medicine will wind up just being explicit about your dependencies and not relying at all on the include_recipe in git_client.  That will always work, and nobody needs to think about it that hard.  While it 'should' work in an ideal world to have git_client be able to express its dependencies, in practice the chef-client parser makes it hard to get right.

This is rather a lot like operator precedence and the use of parens.   Generally the over-use of parens is encouraged in order to be explicit and clear about ordering and resolve any ambiguity in the part of the reader caused by not recalling exactly what the operator precedence happens to be in the language in question.  Similarly here, just do the include_recipe explicitly in your recipe code so nobody needs to think that hard about what happens with include_recipes in LWRPs.

**Comment by lamont-granquist** [comment]
TL;DR:  don't rely on include_recipe in LWRPs/Providers
